File: crawley/parser/javaWorkspace/src/Parser.java
import java.util.*;

public class Parser {

	private static final String CLOSING_PARENTHESIS = "')";
	private static final String PYQUERY_HEAD = "PyQuery(html).query('";
	private static final String RETURN = "return ";
	private static final String ACTION_SEPARATOR = " < ";
	private static final String QUERY_SEPARATOR = " => ";

	public String parse(String crawleyDSL) {
		String[] parsingString = crawleyDSL.split(QUERY_SEPARATOR);
		String[] actionSection = parsingString[0].toLowerCase().split(ACTION_SEPARATOR);
		String action = actionSection[0];
		String properties = actionSection[1];
		// String getSection = parsingString[1];

		return getFinalQuery(action, properties);
	}

	protected String getFinalQuery(String action, String properties) {
		LinkedHashMap<String, String[]> propertyMap = new LinkedHashMap<String, String[]>();

		for (String name : properties.split(" ")) propertyMap.put(name.split(":")[0], Utils.removeBraces(name.split(":")[1]).split(","));

		String result = RETURN + Utils.compoundPropertyStartingBraces(properties);

		for (String property : propertyMap.keySet()) {
			for (int j = 0; j < propertyMap.get(property).length; j++)
				result += ((j == 0) ? "" : ", ") + PYQUERY_HEAD
						+ Property.getProperty(property)
						+ Utils.trimSingleQuotes(propertyMap.get(property)[j])
						+ CLOSING_PARENTHESIS + Action.getAction(action);
		}

		return result + Utils.compoundPropertyEndingBraces(properties);
	}

}

File: crawley/parser/py_parser/parser.py
from actions import *
from properties import *

class Parser(object):

    CLOSING_PARENTHESIS = "')"
    PYQUERY_HEAD = "PyQuery(html).query('"
    RETURN = "return "
    ACTION_SEPARATOR = " < "
    QUERY_SEPARATOR = " => "
    actionElements = {}
    propertyElements = {}

    def __init__(self):
        
        self.populateMaps()

    def populateMaps(self):
        
        self.populateActions()
        self.populateProperties()

    def populateProperties(self):
        
        self.propertyElements["id"] = IDProperty()
        self.propertyElements["tag"] = TagProperty()
        self.propertyElements["class"] = ClassProperty()

    def populateActions(self):
        
        self.actionElements["first"] = FirstAction()
        self.actionElements["last"] = LastAction()
        self.actionElements["all"] = AllAction()

    def parse(self, crawleyDSL):
        
        parsingString = crawleyDSL.split(self.QUERY_SEPARATOR)
        action, properties = parsingString[0].lower().split(self.ACTION_SEPARATOR)
        getSection = parsingString[1]

        return self.getFinalQuery(action, properties)
        
    def getFinalQuery(self, action, properties):
        
        propertyMap = {}

        for name in properties.split(): 
            propertyMap[name.split(":")[0]] = self.removeBraces(name.split(":")[1]).split(",")

        result = self.RETURN + self.compoundPropertyStartingBraces(properties)
        
        for property in propertyMap.keys():
            for j, x in enumerate(propertyMap.get(property)):
                result = "%s%s%s%s%s%s%s" % (result, "" if j == 0 else ", ", self.PYQUERY_HEAD, 
                         self.propertyElements.get(property),
                         self.trimSingleQuotes(propertyMap.get(property)[j]),
                         self.CLOSING_PARENTHESIS,
                         self.actionElements.get(action))
                
        res = "%s%s" % (result, self.compoundPropertyEndingBraces(properties))
        print res
        return res

    def removeBraces(self, propertyValues):
        
        return propertyValues.replace("[", "").replace("]", "")

    def compoundPropertyEndingBraces(self, properties):
        
        return self.compoundPropertyBraces(properties, False)

    def compoundPropertyStartingBraces(self, properties):
        
        return self.compoundPropertyBraces(properties, True)

    def compoundPropertyBraces(self, properties, starting_position):
        
        if "[" in properties and "]" in properties:
            if starting_position:
                return "["
            return "]"
        return ""

    def trimSingleQuotes(self, htmlElement):

        return htmlElement.replace("'", "")
